synapse/replication/tcp/handler.py

```diff
@@ -611,10 +611,14 @@ async def _process_position(
         # Find where we previously streamed up to.
         current_token = stream.current_token(cmd.instance_name)
 
-        # If the position token matches our current token then we're up to
-        # date and there's nothing to do. Otherwise, fetch all updates
-        # between then and now.
-        missing_updates = cmd.prev_token != current_token
+        # If the position token matches our current token then we're up to date
+        # and there's nothing to do. Otherwise, fetch all updates between then
+        # and now.
+        #
+        # Note: We have to check that `current_token` is within the range, to
+        # handle the case where the stream gets "reset" (e.g. for `caches` and
+        # `typing` after the writer's restart).
+        missing_updates = not (cmd.prev_token <= current_token <= cmd.new_token)
```

Comment:
Sanity check: are the inequalities correct here? E.g. should one of them be `<` instead of `<=`?